### DISCUSSION TITLE
Beginner help: port to PHP 8.3

### DISCUSSION THREAD

**Discussion by duboism** [rectorphp/rector#8900] [Q&A] [answered]
Hello,

I'm very much a beginner with Rector so pardon my ignorance.

I have a code that runs on PHP 7.4 but is full of deprecated constructs (for instance `while/each` loops). We want to port it to PHP 8.3 with Rector (if possible keeping compatibility with PHP 7.4).

Inspired by what I found online, I modified `composer.json` like this:
```diff
--- composer.json
+++ composer.json
@@ -4,7 +4,7 @@
         "mpdf/mpdf": "^5.7.4",
         "pear/html_template_it": "^1.3.1",
         "pear/image_canvas": "^0.4.0",
-        "php": " >= 7, < 8",
+        "php": " >= 7",
         "ext-mysqli": "*",
         "ext-mbstring": "*",
         "ext-gd": "*",
```
and created the `rector.php` file with the content:
```php
<?php

declare(strict_types=1);

use Rector\Config\RectorConfig;

use Rector\Set\ValueObject\LevelSetList;

return RectorConfig::configure()
    ->withPaths([
        # Skipped
    ])
    ->withSets([
        LevelSetList::UP_TO_PHP_83,
    ]);
```

Then I tried to convert one file which contain `while/each` loops with: `vendor/bin/rector -vvv --clear-cache process my_file`.

To my surprise, the following rules were applied:
 * LongArrayToShortArrayRector
 * RandomFunctionRector

But the `while/each` loops are still there.

What am I doing wrong ?

I have noticed that if I change the `composer.json` to require `php >= 8`, the `while/each` loops are converted but I don't know why.

Thanks in advance !

P.S.: I have found similar discussions but could make sense of them:
  - #8285
  - #8786

**Comment by duboism** [comment]
Any help?

**Comment by darthf1** [comment, marked as answer]
I think it's all here:

https://getrector.com/documentation/integration-to-new-project

**Reply by duboism** [reply]
Hello,

Thanks for your answer. I found this document a few days ago but I did not understand it totally and didn't have time to experiment. So if I summarize, it's an iterative process where you put a version of PHP inside `withPhpSets`, update, switch  to the next version and so on. Correct ?

My second concern is that I don't really care about rules like `LongArrayToShortArrayRector` (it's a lot of changes for something that is works). For now, I want to focus on breaking changes (typically `while/each` loops). Is there a way to do that ?

I found another approach (https://github.com/rectorphp/rector/pull/3287#issuecomment-641567656) where rules are applied 1 by 1: you find a problem in yout code base, apply the rule to correct it... But it seems that this post dates back to the old YAML config file.
